File: src/main/java/bad/robot/radiate/monitor/MonitoringTasksFactory.java
package bad.robot.radiate.monitor;

import bad.robot.radiate.teamcity.AllProjectsAsSingleTask;
import bad.robot.radiate.teamcity.AllProjectsOneTaskPerProject;

import java.util.List;

@Deprecated
public interface MonitoringTasksFactory extends Observable {

    List<MonitoringTask> create();

    /** Default mode */
    static MonitoringTasksFactory singleAggregate() {
        return new AllProjectsAsSingleTask();
    }

    /** Chessboard mode */
    static MonitoringTasksFactory multipleProjects() {
        return new AllProjectsOneTaskPerProject();
    }

    static MonitoringTasksFactory multipleBuildsDemo() {
        return new MultipleBuildsDemo();
    }

    static MonitoringTasksFactory demo() {
        return new Demo();
    }

    static MonitoringTasksFactory erroring() {
        return new Error();
    }

    class Error extends ThreadSafeObservable implements MonitoringTasksFactory {
        public List<MonitoringTask> create() {
            throw new RuntimeException("An unrecoverable error occurred");
        }
    }

}

File: src/main/scala/bad/robot/radiate/monitor/MonitoringTasksFactory.scala
package bad.robot.radiate.monitor

import bad.robot.radiate.ParseError
import bad.robot.radiate.teamcity.{AllProjectsAsSingleTask, AllProjectsOneTaskPerProject}
import bad.robot.radiate.Error
import bad.robot.radiate.travis.TravisProjects
import scalaz.\/

trait MonitoringTasksFactory extends Observable {
  def create: Error \/ List[MonitoringTask]
}

object MonitoringTasksFactory {
  /** Default mode */
  def singleAggregate = new AllProjectsAsSingleTask

  /** Chessboard mode */
  def multipleProjects = new AllProjectsOneTaskPerProject

  def multipleBuildsDemo = new MultipleBuildsDemo
  
  def demo = new Demo

  def travis = new TravisProjects

  def erroring = new ThreadSafeObservable with MonitoringTasksFactory {
    import scalaz.syntax.either._
    def create = ParseError("An unrecoverable error occurred").left
  }

}
